internal/backend/db/dbgorm/tx.go

```diff
@@ -10,59 +10,32 @@ import (
 	"go.autokitteh.dev/autokitteh/internal/backend/db"
 )
 
-type tx struct {
-	gormdb
-	ctx  context.Context
-	done bool
-}
-
-func (db *gormdb) Begin(ctx context.Context) (db.Transaction, error) {
-	db.db.Error = nil
-	tx1 := db.db.WithContext(ctx).Begin()
-	if err := tx1.Error; err != nil {
-		return nil, err
-	}
-
-	return &tx{
-		gormdb: gormdb{
-			z:  db.z.With(zap.String("txid", uuid.NewString())),
-			db: tx1,
-		},
-	}, nil
-}
-
-func (tx *tx) Commit() error {
-	tx.done = true
-	return tx.db.Commit().Error
-}
-
-func (tx *tx) Rollback() error {
-	if tx.done {
-		return nil
-	}
-
-	return tx.db.Rollback().Error
+func (db *gormdb) Transaction(ctx context.Context, f func(db db.DB) error) error {
+	return db.transaction(ctx, func(tx *gormdb) error { return f(tx) })
 }
 
-func (db *gormdb) Transaction(ctx context.Context, f func(db db.DB) error) error {
-	return db.transaction(ctx, func(tx *tx) error {
-		return f(tx)
+func (db *gormdb) transaction(ctx context.Context, f func(tx *gormdb) error) error {
+	return db.wdb.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
+		return f(
+			&gormdb{
+				z:   db.z.With(zap.String("txid", uuid.NewString())),
+				wdb: tx,
+				rdb: tx,
+				cfg: db.cfg,
+			},
+		)
 	})
 }
 
-func (db *gormdb) transaction(ctx context.Context, f func(tx *tx) error) error {
-	return db.locked(func(db *gormdb) error {
-		return db.db.WithContext(ctx).Transaction(func(txdb *gorm.DB) error {
-			return f(
-				&tx{
-					gormdb: gormdb{
-						z:   db.z.With(zap.String("txid", uuid.NewString())),
-						db:  txdb,
-						cfg: db.cfg,
-					},
-					ctx: ctx,
-				},
-			)
-		})
+func (db *gormdb) rtransaction(ctx context.Context, f func(tx *gormdb) error) error {
+	return db.rdb.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
+		return f(
+			&gormdb{
+				z:   db.z.With(zap.String("txid", uuid.NewString())),
+				wdb: nil, // panic on writes.
```

Comment:
This is a footgun. A panic is not an appropriate result for mistakenly using a read-only connection to the DB. AK should return errors on such transaction, and the code using transactions should handle errors.

Reply:
tbh i don't know how to prevent this. the code itself should not do any writes in this case and i think it warrants a panic. in any case, how this can be solved?